demos/qml_optimization.ipynb

```diff
@@ -0,0 +1,469 @@
+        "\n",
+        "In this example, we'll show you how to use [JAX](https://jax.readthedocs.io), an autodifferentiable machine learning framework, and [Optax](https://optax.readthedocs.io/), a suite of JAX-compatible gradient-based optimizers, to optimize a PennyLane quantum machine learning model.\n",
+        "\n",
+        "## Set up your model, data, and cost\n",
+        "\n",
+        "Here, we will create a simple QML model for our optimization. In particular:\n",
+        "\n",
+        "- We will embed our data through a series of rotation gates.\n",
+        "- We will then have an ansatz of trainable rotation gates with parameters `weights`; it is these values we will train to minimize our cost function.\n",
+        "- We will train the QML model on `data`, a `(5, 4)` array, and optimize the model to match target predictions given by `target`."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 3,
+      "metadata": {
+        "id": "iKtDMXjENR7X"
+      },
+      "outputs": [],
+      "source": [
+        "import pennylane as qml\n",
+        "from jax import numpy as jnp\n",
+        "import jax\n",
+        "import optax\n",
+        "import catalyst"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 4,
+      "metadata": {
+        "id": "tiK-itpFNVwE"
+      },
+      "outputs": [],
+      "source": [
+        "n_wires = 5\n",
+        "data = jnp.sin(jnp.mgrid[-2:2:0.2].reshape(n_wires, -1)) ** 3\n",
+        "targets = jnp.array([-0.2, 0.4, 0.35, 0.2])\n",
+        "\n",
+        "dev = qml.device(\"lightning.qubit\", wires=n_wires)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 5,
+      "metadata": {
+        "id": "w4QPSXdysqJx"
+      },
+      "outputs": [],
+      "source": [
+        "@qml.qnode(dev)\n",
+        "def circuit(data, weights):\n",
+        "    \"\"\"Quantum circuit ansatz\"\"\"\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
+        "    def data_embedding(i):\n",
+        "        qml.RY(data[i], wires=i)\n",
+        "\n",
+        "    data_embedding()\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
+        "    def ansatz(i):\n",
+        "        qml.RX(weights[i, 0], wires=i)\n",
+        "        qml.RY(weights[i, 1], wires=i)\n",
+        "        qml.RX(weights[i, 2], wires=i)\n",
+        "        qml.CNOT(wires=[i, (i + 1) % n_wires])\n",
+        "\n",
+        "    ansatz()\n",
+        "\n",
+        "    # we use a sum of local Z's as an observable since a\n",
+        "    # local Z would only be affected by params on that qubit.\n",
+        "    return qml.expval(qml.sum(*[qml.PauliZ(i) for i in range(n_wires)]))\n",
+        "\n",
+        "circuit = qml.qjit(catalyst.vmap(circuit, in_axes=(1, None)), autograph=False)"
```

Comment:
Is the batching on axis 1 intentional? (I think it will create some partial transposition compared to the "intuitive" result of vmapping)

Reply:
Batching axis of 1 is intentional, because I'm matching the results of this demo here to make sure I get the same results:

https://pennylane.ai/qml/demos/tutorial_How_to_optimize_QML_model_using_JAX_and_Optax/

Here, vmap isn't used; instead, operator vectorization is used directly:

![image](https://github.com/PennyLaneAI/catalyst/assets/2959003/25132ade-f9d2-432e-abc5-22de22feece7)

The interesting thing about operator vectorization is that it is roughly equivalent to a batch dimension at the _end_ of the data, rather than the beginning! Because it's the 'last' axes of the data array that is provided to the operator.

Reply:
Oh interesting, good to know! To readers the `in_axes=(1, None)` might be coming out of the blue though.